### PR TITLE
[gh-518] add Schnorr validator over Secp256k1.

### LINKED ISSUE #514: [secp256k1] support verification logics for ECDSA and Schnorr signatures
Like Ed25519 currently implemented for validators, there should also be implementations for verification on validators based on the design of https://github.com/rooch-network/rooch/issues/460 and the implementation of https://github.com/rooch-network/rooch/pull/466.

At this time, there's no verification implemented for ECDSA and Schnorr signatures, only dummy remains.

This issue can be further divided to:

1. #518 
2. #519 
3. #549

### LINKED ISSUE #518: [validator] add a Schnorr validator
Part of #514

### PR DESCRIPTION
**Pull request by feliciss**
Resolves #518

### CODE REVIEW

**Comment by feliciss** [comment]
I have encountered an error and am still striving to resolve it.

```rs
thread 'tests::schnorr_validator_tests::test_validate' panicked at 'called `Result::unwrap()` on an `Err` value: VMError with status VECTOR_OPERATION_ERROR with sub status 1 at location Module ModuleId { address: 0000000000000000000000000000000000000000000000000000000000000003, name: Identifier("schnorr_validator") } at code offset 17 in function definition 6', crates/rooch-framework-tests/src/tests/schnorr_validator_tests.rs:29:10
```

If I make

```move
const SCHNORR_PUBKEY_LENGTH: u64 = 64;
```

shorter to 32, just like the Ed25519 public key, there will be an another error:

```rs
thread 'tests::schnorr_validator_tests::test_validate' panicked at 'called `Result::unwrap()` on an `Err` value: VMError with status ABORTED with sub status 66537 at location Module ModuleId { address: 0000000000000000000000000000000000000000000000000000000000000003, name: Identifier("schnorr_validator") } and message 0x0000000000000000000000000000000000000000000000000000000000000003::schnorr_validator::validate at offset 14 at code offset 14 in function definition 9', crates/rooch-framework-tests/src/tests/schnorr_validator_tests.rs:29:10
stack backtrace:
```

Since I'm not very familiar with moveosvm, can you @jolestar check the file https://github.com/rooch-network/rooch/pull/517/files#diff-1d9159dbfcd0de959bf1f8d6f8ce44189d6cff88c74cabbfd6ff2163bc69b34f to see if anything went wrong?

File: crates/rooch-framework/sources/auth_validator/schnorr_validator.move [hunk 1]
```diff
@@ -14,15 +22,83 @@ module rooch_framework::schnorr_validator {
...
+   }
+
    public fun validate(ctx: &StorageContext, payload: vector<u8>){
-      //FIXME check the address and public key relationship
+      let auth_key = schnorr_authentication_key(&payload);
+      let auth_key_in_account = get_authentication_key(ctx, storage_context::sender(ctx));
+      assert!(
+         auth_key_in_account == auth_key,
+         auth_validator::error_invalid_account_auth_key()
+      );
```
Comment (jolestar) [inline crates/rooch-framework/sources/auth_validator/schnorr_validator.move:92]: It seam like the error from here.

File: crates/rooch-framework/sources/auth_validator/schnorr_validator.move [hunk 2]
```diff
@@ -14,15 +22,83 @@ module rooch_framework::schnorr_validator {
...
+      moveos_std::bcs::to_address(hash::blake2b256(&bytes))
+   }
+
+   public fun get_authentication_key(ctx: &StorageContext, addr: address): vector<u8> {
+      let auth_key_option = account_authentication::get_authentication_key<SchnorrValidator>(ctx, addr);
+      if(option::is_some(&auth_key_option)){
+         option::extract(&mut auth_key_option)
+      }else{
+        //if AuthenticationKey does not exist, return addr as authentication key
+        moveos_std::bcs::to_bytes(&addr)
```
Comment (jolestar) [inline crates/rooch-framework/sources/auth_validator/schnorr_validator.move:82]: The address is not a Shnorr public key's hash, so we can not use the address as authentication_key.

File: crates/rooch-framework/sources/auth_validator/schnorr_validator.move (hunk 1 above)
Comment (jolestar) [reply crates/rooch-framework/sources/auth_validator/schnorr_validator.move:92]: You can skp the account auth key here. We need to figure out how to handle no ed25519's auth key.

Comment (feliciss) [reply crates/rooch-framework/sources/auth_validator/schnorr_validator.move:92]: I skipped the auth key comparison. Now it comes a new error:

```zsh
thread 'tests::schnorr_validator_tests::test_validate' panicked at 'called `Result::unwrap()` on an `Err` value: VMError with status ABORTED with sub status 1 at location Module ModuleId { address: 0000000000000000000000000000000000000000000000000000000000000003, name: Identifier("schnorr") } at code offset 0 in function definition 0', crates/rooch-framework-tests/src/tests/schnorr_validator_tests.rs:29:10
```

I don't really know what the `ABORTED with sub status 1` means.

File: crates/rooch-framework/sources/auth_validator/schnorr_validator.move (hunk 2 above)
Comment (feliciss) [reply crates/rooch-framework/sources/auth_validator/schnorr_validator.move:82]: Yes. May need to support other non-ed25519's auth keys as built-in schemes.

File: crates/rooch-framework/sources/auth_validator/schnorr_validator.move (hunk 1 above)
Comment (feliciss) [reply crates/rooch-framework/sources/auth_validator/schnorr_validator.move:92]: The code `1` might be related to the code defined as in `schnorr.rs`?

```rs
pub const INVALID_PUBKEY: u64 = 1;
```

Let me try to resolve it.

File: crates/rooch-framework/sources/auth_validator/schnorr_validator.move (hunk 2 above)
Comment (jolestar) [reply crates/rooch-framework/sources/auth_validator/schnorr_validator.move:82]: non-ed25519's auth keys need to be inited first, then the validator uses the auth keys.

**Comment by jolestar** [comment]
Could you pass the test(skip some conditions)? And We can merge this PR and #524.  I will do some follow refactoring.

**Comment by feliciss** [comment]
> Could you pass the test(skip some conditions)? And We can merge this PR and #524. I will do some follow refactoring.

Resolved. You can review it now.

File: crates/rooch-framework-tests/src/tests/schnorr_validator_tests.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use moveos_types::transaction::MoveAction;
+use rooch_framework::bindings::empty::Empty;
+use rooch_key::keystore::{AccountKeystore, InMemKeystore};
+use rooch_types::transaction::{rooch::RoochTransactionData, AbstractTransaction};
+
+use crate::binding_test;
+
+#[ignore = "https://github.com/rooch-network/rooch/pull/517#issuecomment-1648866482"]
```
Comment (feliciss) [inline crates/rooch-framework-tests/src/tests/schnorr_validator_tests.rs:11]: Ignored.

**Review by jolestar — APPROVED** [review]

**Pull request merged**